lib/core/src/Cardano/Wallet/DB/Model.hs

```diff
@@ -0,0 +1,235 @@
+
+mListWallets :: Ord wid => ModelOp wid s t xprv [wid]
+mListWallets db@(Database wallets _) = (Right (sort $ Map.keys wallets), db)
+
+mPutCheckpoint :: Ord wid => wid -> Wallet s t -> ModelOp wid s t xprv ()
+mPutCheckpoint wid cp = alterModel wid $ \wal ->
+    ((), wal { checkpoints = [cp] })
+
+mReadCheckpoint :: Ord wid => wid -> ModelOp wid s t xprv (Maybe (Wallet s t))
+mReadCheckpoint wid db@(Database wallets _) =
+    (Right (checkpoints <$> Map.lookup wid wallets >>= listToMaybe), db)
+
+mPutWalletMeta :: Ord wid => wid -> WalletMetadata -> ModelOp wid s t xprv ()
+mPutWalletMeta wid meta = alterModel wid $ \wal ->
+    ((), wal { metadata = meta })
+
+mReadWalletMeta :: Ord wid => wid -> ModelOp wid s t xprv (Maybe WalletMetadata)
+mReadWalletMeta wid db@(Database wallets _) =
+    (Right (metadata <$> Map.lookup wid wallets), db)
+
+mPutTxHistory :: Ord wid => wid -> TxHistoryMap t -> ModelOp wid s t xprv ()
+mPutTxHistory wid txList db@Database{wallets,txs} =
+    case Map.lookup wid wallets of
+        Just wal -> (Right (), Database (Map.insert wid wal' wallets) txs')
+          where
+            wal' = wal { txHistory = txHistory wal <> Map.map snd txList }
+            txs' = txs <> Map.map fst txList
+        Nothing -> (Left (NoSuchWallet wid), db)
+
+mReadTxHistory
+    :: forall wid s t xprv.
+       (Ord wid)
+    => wid
+    -> SortOrder
+    -> Range SlotId
+    -> ModelOp wid s t xprv (TxHistory t)
+mReadTxHistory wid order range db@Database{wallets,txs} = (Right res, db)
+  where
+    res = maybe mempty getTxs $ Map.lookup wid wallets
+    getTxs wal = filterTxHistory @t order range $ catMaybes
+            [ fmap (\tx -> (txId, (tx, meta))) (Map.lookup txId txs)
+            | (txId, meta) <- Map.toList (txHistory wal) ]
+
+mPutPrivateKey :: Ord wid => wid -> xprv -> ModelOp wid s t xprv ()
+mPutPrivateKey wid pk = alterModel wid $ \wal ->
+    ((), wal { xprv = Just pk })
+
+mReadPrivateKey :: Ord wid => wid -> ModelOp wid s t xprv (Maybe xprv)
+mReadPrivateKey wid db@(Database wallets _) =
+    (Right (Map.lookup wid wallets >>= xprv), db)
+
+{-------------------------------------------------------------------------------
+                             Model function helpers
+-------------------------------------------------------------------------------}
+
+alterModel
+    :: Ord wid
+    => wid
+    -> (WalletDatabase s t xprv -> (a, WalletDatabase s t xprv))
+    -> ModelOp wid s t xprv a
+alterModel wid f db@Database{wallets,txs} = case f <$> Map.lookup wid wallets of
+    Just (a, wal) -> (Right a, Database (Map.insert wid wal wallets) txs)
+    Nothing -> (Left (NoSuchWallet wid), db)
+
+-- | Apply optional filters on slotId and sort using the default sort order
+-- (first time/slotId, then by TxId) to a 'TxHistory'.
+filterTxHistory :: SortOrder -> Range SlotId -> TxHistory t -> TxHistory t
+filterTxHistory order range =
+    filter ((`isWithinRange` range) . slotId . snd . snd)
+    . (case order of
```

Comment:
We _could_ use `sortBy` + `Down` instead of `sortOn` + `reverse` to traverse the list one time less, if it matters for performance.

Reply:
I wanted this code to be as simple and obvious as possible given that inefficiency doesn't matter in this case (particularly not linear slowdowns). So when I cut & pasted this function I was happy to leave it as it was.